src/auth/auth.service.ts

```diff
@@ -112,4 +130,60 @@ export class AuthService {
       password: hashedPassword,
     });
   }
+
+  async createQrcodeFor2FA(userId: number): Promise<string> {
+    const email = (await this.userRepository.getUserById(userId)).email;
+    const service = 'Recipe App';
+
+    const otpauth = authenticator.keyuri(
+      email,
+      service,
+      process.env.SECRET_KEY_2FA,
+    );
+
+    return qrcode.toDataURL(otpauth);
+  }
+
+  async enable2FA(userId: number, providedToken: string): Promise<string[]> {
+    const user = await this.userRepository.getUserById(userId);
+
+    if (authenticator.check(providedToken, process.env.SECRET_KEY_2FA)) {
+      const recoveryKeys: string[] = Array.from({ length: 3 }, () =>
+        authenticator.generateSecret(),
+      );
+      this.userRepository.enable2FAForUserWithId(user.id, recoveryKeys);
+      return recoveryKeys;
+    } else {
+      throw new UnauthorizedException('Incorrect 2FA token');
+    }
+  }
+
+  async disable2FA(userId: number): Promise<UserPayloadRequest> {
+    return this.userRepository.disable2FAForUserWithId(userId);
+  }
+
+  async verify2FA(userId: number, token: string): Promise<string> {
+    const user = await this.userRepository.getUserById(userId);
+    const secret = process.env.SECRET_KEY_2FA;
+
+    if (authenticator.check(token, secret)) {
+      return this.successfullLoginToken(user.id, user.email);
+    } else {
+      throw new UnauthorizedException('Incorrect 2FA token');
+    }
+  }
+
+  async recoverAccountWith2FA(
+    userId: number,
+    providedRecoveryKey: string,
+  ): Promise<string> {
+    const { recoveryKeys, email } =
+      await this.userRepository.get2FARecoveryKeysAndEmailByUserId(userId);
+
+    if (!recoveryKeys.includes(providedRecoveryKey)) {
+      throw new ForbiddenException();
+    }
+
+    return this.successfullLoginToken(userId, email);
+  }
```

Comment:
After the recovery code is used, it should be marked as used and should no longer be available for use. Consider also an endpoint for regenerating the recovery codes.